[Do not review] Add custom scheduler in presto-router

pdabre12 opened the pull request:
## Description
<!---Describe your changes in detail-->

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* ... 
* ... 

Hive Connector Changes
* ... 
* ... 
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

pdabre12 commented:
CC : @aaneja

tdcmeehan commented on presto-spi/pom.xml:
```diff
@@ -53,6 +53,18 @@
             <artifactId>jol-core</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
We do not want to add these dependencies to the SPI.  Guava in particular is heavily used and likely to cause conflicts on the developers who need to bring it in.  Please refactor to avoid these dependencies.

tdcmeehan replied:
I would keep `RequestInfo` where it is, but create a new DTO in the SPI that we can translate to/from.

ZacBlanco reviewed (changes requested):
For the most part this change is ok, but this is a large PR. I think the change in the router SPI to accept `RequestInfo` on `getDestination` should be a separate PR though. (1) it minimizes the diff and (2) is different enough in scope from the plan checker that I don't think it should be rolled into this PR.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginConfig.java:
```diff
@@ -0,0 +1,73 @@
+    private static final Splitter SPLITTER = Splitter.on(',').trimResults().omitEmptyStrings();
+
+    private List<URI> planCheckClustersURIs;
+    private URI javaRouterURI;
+    private URI nativeRouterURI;
```
We are routing to other routers.

ZacBlanco replied:
I didn't know we had the ability to do that. Neat!

pdabre12 commented on presto-native-sidecar-plugin/src/test/java/com/facebook/presto/sidecar/TestNativeSidecarPlanChecker.java:
```diff
@@ -0,0 +1,84 @@
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        DistributedQueryRunner queryRunner = (DistributedQueryRunner) PrestoNativeQueryRunnerUtils.createQueryRunner(true, true, false, false, ImmutableMap.of("http-server.http.port", "8089"));
```
The changes are part of the cherry-pick of https://github.com/prestodb/presto/pull/25115

pdabre12 commented on presto-native-sidecar-plugin/src/main/java/com/facebook/presto/sidecar/nativechecker/NativePlanChecker.java:
```diff
@@ -70,12 +83,33 @@ public void validate(PlanNode planNode, WarningCollector warningCollector, Conne
+    }
+
+    /**
+     * HACK: Replace TableWriterNode from the plan fragment with a ProjectNode because validating a TableWriterNode
```
I am not really sure, @aditi-pandit maybe you can help address this comment?

aditi-pandit replied:
@ZacBlanco : It is supported in native, but there are some fields only populated by the scheduler... Ideally we should have some validation minus those properties. We can do that in the future.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,243 @@
+
+    public PlanCheckerRouterPluginPrestoClient(URI preonClusterURI, URI javaRouterURI, URI nativeRouterURI)
+    {
+        this.preonClusterURI = preonClusterURI;
```
Removed.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginConfig.java:
```diff
@@ -0,0 +1,73 @@
+    private URI javaRouterURI;
+    private URI nativeRouterURI;
+
+    @Config("plan-check-clusters-uris")
```
Done.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginConfig.java:
```diff
@@ -0,0 +1,73 @@
+            return this;
+        }
+
+        this.planCheckClustersURIs = ImmutableList.copyOf(transform(SPLITTER.split(uris), URI::create));
```
`SPLITTER.split` returns an `Iterable<String>` which `Arrays.stream()` don't seem to operate on, unless you want me to not use `SPLITTER` at all?

steveburnett commented on presto-plan-checker-router-plugin/README.md:
```diff
@@ -0,0 +1,32 @@
+Place the plugin jar and all the dependent jars for the plugin in the `plugin` directory relative to the Presto install directory.
+
+The configuration file for this plugin must be `etc/router-config/router-scheduler.properties`.  
+In `router-scheduler.properties`, set the name of the custom scheduler factory.  
```
```suggestion
Use `router-scheduler.properties`to set the name of the custom scheduler factory.  
```

steveburnett commented on presto-plan-checker-router-plugin/README.md:
```diff
@@ -0,0 +1,32 @@
+
+The configuration file for this plugin must be `etc/router-config/router-scheduler.properties`.  
+In `router-scheduler.properties`, set the name of the custom scheduler factory.  
+For example, use the following line to set the custom scheduler factory name to `plan-checker`.  
```
```suggestion
For example, set the custom scheduler factory name to `plan-checker`:  
```

steveburnett commented on presto-plan-checker-router-plugin/README.md:
```diff
@@ -0,0 +1,32 @@
+  This class implements the interface `com.facebook.presto.spi.SchedulerFactory`.
+* `MetricsBasedScheduler` - Example custom scheduler implementing the scheduling logic for clusters.  
+  This class implements the interface `com.facebook.presto.spi.router.Scheduler`.  
+  Similar classes can be added implementing specific custom scheduling logic.
```
```suggestion

Similar classes can be added implementing specific custom scheduling logic.
```
Suggest adding a new line and unindenting this sentence: the way it's currently formatted it's within the `MetricsBasedScheduler` list item and that formatting suggests this sentence applies to that, not the general statement I think you intend this to be.

pdabre12 replied:
Removed this line.

steveburnett commented on presto-plan-checker-router-plugin/README.md:
```diff
@@ -0,0 +1,32 @@
+  Similar classes can be added implementing specific custom scheduling logic.
+
+## Configuration:
+The following config properties are required to be set:
```
```suggestion
The following configuration properties must be set in `etc/router-config/router-scheduler.properties`:
```

steveburnett reviewed (changes requested):
Thanks for the doc! 

In addition to the line suggestions I made here, I thought that maybe you could move the one config instruction out of the first part down into the Configuration section. Line editing for this last suggestion got too complicated, so I wrote a mockup and here's a screenshot of it. Take a look and let me know what you think. 

<img width="753" alt="Screenshot 2025-06-03 at 10 27 33 AM" src="https://github.com/user-attachments/assets/23658c0e-b524-4fc8-8ee3-64f797d8b060" />

pdabre12 commented:
@steveburnett Thanks for the suggestions. 
I addressed your comments, PTAL.

steveburnett commented on presto-plan-checker-router-plugin/README.md:
```diff
@@ -0,0 +1,31 @@
+## Add the Custom Scheduler Plugin to the Presto Router
+Place the plugin jar and all the dependent jars for the plugin in the `plugin` directory relative to the Presto install directory.
+
+Create a configuration file for this plugin must be `etc/router-config/router-scheduler.properties`.  
```
```suggestion
Create a configuration file for this plugin. The file must be named 'router-scheduler.properties` and must be in the `etc/router-config/` directory relative to the Presto install directory.  
```
There's several ways you could fix this line, this is just one suggestion. If this one doesn't work for you, find another one that works better for you.

steveburnett reviewed (changes requested):
Thanks for the doc changes! Found one more item that should be changed.

pdabre12 commented on presto-spi/src/main/java/com/facebook/presto/spi/router/Scheduler.java:
```diff
@@ -24,17 +24,7 @@ public interface Scheduler
-    {
-        return Optional.empty();
-    };
+    Optional<URI> getDestination(RequestInfo requestInfo);
```
Opened the change to `RequestInfo` in new PR here: https://github.com/prestodb/presto/pull/25244

aditi-pandit reviewed (commented):
@pdabre12 : Looked at your code at a high level as I'm not very familiar with the classes, and it seems fine.

Only had a question about testing though : Have you tested with the SQL in presto-native-tests (atleast enabling the plan checker) ?

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,244 @@
+            verify(client.isFinished());
+            QueryError resultsError = client.finalStatusInfo().getError();
+            if (resultsError != null) {
+                if (ErrorType.valueOf(resultsError.getErrorType()) == USER_ERROR) {
```
```
if (resultsError != null) {
                if (ErrorType.valueOf(resultsError.getErrorType()) == USER_ERROR) {
                    throw new PrestoException(
                            () -> new ErrorCode(
                                    resultsError.getErrorCode(),
                                    resultsError.getErrorName(),
                                    USER_ERROR,
                                    resultsError.isRetriable()),
                            newQueryResults(client));
                }
                isNativeCompatible = false;
                log.info(resultsError.getMessage());
}
```

We are actually checking for user errors and sending it back to the presto client from which the query came from assuming this would fail for both clusters.

If its not an `USER_ERROR` , only then we tag it as native-incompatible and send it over to the java cluster.

tdcmeehan replied:
Oh, you're right--I misread that.

pdabre12 commented on .github/workflows/prestocpp-linux-build-and-unit-test.yml:
```diff
@@ -309,3 +309,72 @@ jobs:
+      - name: Run presto-native plan checker router plugin tests
+        run: |
+          export PRESTO_SERVER_PATH="${GITHUB_WORKSPACE}/presto-native-execution/_build/release/presto_cpp/main/presto_server"
+          export TESTFILES=`find ./presto-plan-checker-router-plugin/src/test -type f -name 'Test*.java'`
+          # Convert file paths to comma separated class names
+          export TESTCLASSES=
+          for test_file in $TESTFILES
+          do
+            tmp=${test_file##*/}
+            test_class=${tmp%%\.*}
+            export TESTCLASSES="${TESTCLASSES},$test_class"
+          done
+          export TESTCLASSES=${TESTCLASSES#,}
+          echo "TESTCLASSES = $TESTCLASSES"
```
Sounds like a good idea, but do you also want me to introduce a profile for the other jobs as well?
If that's the case, can I do this as a followup?

ZacBlanco replied:
Just fix it for this PR and you can fix the other ones later

pdabre12 replied:
Done, PTAL.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginScheduler.java:
```diff
@@ -0,0 +1,86 @@
+    private URI getValidatorDestination(String user)
+    {
+        try {
+            synchronized (planCheckerClusterCandidateIndex) {
```
Done.

pdabre12 commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,244 @@
+
+    private ClientSession parseHeadersToClientSession(HttpServletRequest httpServletRequest)
+    {
+        // todo: How to parse headers into a ClientSession object?
```
Removed this.

pdabre12 commented:
> @pdabre12 : Looked at your code at a high level as I'm not very familiar with the classes, and it seems fine.
> 
> Only had a question about testing though : Have you tested with the SQL in presto-native-tests (atleast enabling the plan checker) ?

@aditi-pandit Yes, we have enabled the sidecar and the native plan checker in `presto-native-tests` module.

steveburnett reviewed (approved):
LGTM! (docs)

aaneja commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,243 @@
+
+        if (isNativeCompatible) {
+            log.debug("Native compatible, routing to native-clusters router: [%s]", nativeRouterURI);
+            nativeClusterRedirectRequests.update(1);
```
nit : `nativeClusterRedirectRequests.update(1L)` you can use a native long constant

aaneja commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,243 @@
+            return Optional.of(nativeRouterURI);
+        }
+        log.debug("Native incompatible, routing to java-clusters router: [%s]", javaRouterURI);
+        javaClusterRedirectRequests.update(1);
```
nit: same here

aaneja commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,243 @@
+                sessionContext.getCatalog(),
+                sessionContext.getSchema(),
+                sessionContext.getTimeZoneId(),
+                Locale.ENGLISH,
```
Can we not derive the original requests LOCALE ? Or does it not matter for EXPLAIN queries ?

pdabre12 replied:
Deriving this from the original request now.

aaneja commented on presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java:
```diff
@@ -0,0 +1,243 @@
+                sessionContext.getIdentity().getRoles(),
+                sessionContext.getIdentity().getExtraCredentials(),
+                httpServletRequest.getHeader(PRESTO_TRANSACTION_ID),
+                new Duration(2, MINUTES),
```
Can we make the the client session timeout configurable ? Users may not want to wait 2 minutes; they may want to fail early

aaneja commented on presto-plan-checker-router-plugin/src/test/java/com/facebook/presto/router/scheduler/TestPlanCheckerRouterPlugin.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Override
+    protected void createTables()
```
I'm missing something I think - are we actually querying these tables ? The test actually only check that the destination the router plugin resolves to..

aaneja commented on presto-plan-checker-router-plugin/src/test/java/com/facebook/presto/router/scheduler/TestPlanCheckerRouterPlugin.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Override
+    protected QueryRunner createExpectedQueryRunner()
```
Similarly, do we need this expected runner to be setup ?

pdabre12 replied:
Don't need to setup this, removed .Thanks.

aaneja commented on presto-router/src/main/java/com/facebook/presto/router/RouterResource.java:
```diff
@@ -60,7 +60,16 @@ public RouterResource(ClusterManager clusterManager)
+            coordinatorUri = clusterManager.getDestination(requestInfo).orElseThrow(() -> badRequest(BAD_GATEWAY, "No Presto cluster available"));
+        }
+        catch (RuntimeException e) {
+            return Response.status(Response.Status.OK)
```
IMO, lets use `Response.Status.INTERNAL_SERVER_ERROR` for this case

aaneja commented on presto-router-example-plugin-scheduler/pom.xml:
```diff
@@ -37,6 +37,26 @@
             <scope>test</scope>
         </dependency>
+
+        <dependency>
```
No related code changes, so these can be undone ?

steveburnett commented:
> Thanks @steveburnett , made the changes, can you PTAL?

Thanks for updating the link! Could you change the section heading `General Changes` to `Router Changes`?

steveburnett commented:
> @steveburnett Done, thanks !

I'm sorry if I was unclear! We've added a new section heading `Router Changes` that this note should go into, instead of the `General Changes` section heading. Apologies for this nit taking so much of your time. 

```
== RELEASE NOTES ==

Router Changes
* Add a new custom router scheduler plugin, the `Presto Plan Checker Router Scheduler Plugin <https://github.com/prestodb/presto/tree/master/presto-plan-checker-router-plugin/README.md>`_.
```

pdabre12 commented:
> > @steveburnett Done, thanks !
> 
> I'm sorry if I was unclear! We've added a new section heading `Router Changes` that this note should go into, instead of the `General Changes` section heading. Apologies for this nit taking so much of your time.
> 
> ```
> == RELEASE NOTES ==
> 
> Router Changes
> * Add a new custom router scheduler plugin, the `Presto Plan Checker Router Scheduler Plugin <https://github.com/prestodb/presto/tree/master/presto-plan-checker-router-plugin/README.md>`_.
> ```

@steveburnett My bad, I mis-read your suggestion , I have moved it to `Router Changes` now.

aditi-pandit reviewed (approved):
Thanks @pdabre12.

We need to investigate these failures the e2e test failures but I don't think they are related to this work https://github.com/prestodb/presto/actions/runs/15566344717/job/43839187753?pr=25035

pdabre12 commented:
> Thanks @pdabre12.
> 
> We need to investigate these failures the e2e test failures but I don't think they are related to this work https://github.com/prestodb/presto/actions/runs/15566344717/job/43839187753?pr=25035

Yes, these failures were seen earlier too: https://github.com/prestodb/presto/actions/runs/15566813960/job/43833560154
